src/System.Text.Json/tests/JsonObjectTests.cs

```diff
@@ -0,0 +1,594 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Xunit;
+
+namespace System.Text.Json.Tests
+{
+    public static partial class JsonObjectTests
+    {
+        [Fact]
+        public static void TestDefaultConstructor()
+        {
+            var jsonObject = new JsonObject();
+            Assert.Equal(0, jsonObject.PropertyNames.Count);
+            Assert.Equal(0, jsonObject.PropertyValues.Count);
+        }
+
+        [Fact]
+        public static void TestIEnumerableKVPConstructor()
+        {
+            var jsonProperties = new List<KeyValuePair<string, JsonNode>>();
+            jsonProperties.Add(new KeyValuePair<string, JsonNode>("number", new JsonNumber(17)));
+            jsonProperties.Add(new KeyValuePair<string, JsonNode>("string", new JsonString("property value")));
+            jsonProperties.Add(new KeyValuePair<string, JsonNode>("boolean", new JsonBoolean(true)));
+
+            var jsonObject = new JsonObject(jsonProperties);
+
+            Assert.Equal(17, ((JsonNumber)jsonObject["number"]).GetInt32());
+            Assert.Equal("property value", ((JsonString)jsonObject["string"]).Value);
+            Assert.True(((JsonBoolean)jsonObject["boolean"]).Value);
+        }
+
+        public enum ExpectedValue
+        {
+            Previous,
+            New
+        }
+
+        private static void TestDuplicates(DuplicatePropertyNameHandling duplicatePropertyNameHandling, ExpectedValue expected = default, bool valueConstructor = true)
+        {
+            string previousValue = "value1";
+            string newValue = "value2";
+
+            JsonObject jsonObject = valueConstructor ? new JsonObject(duplicatePropertyNameHandling) : new JsonObject();
+            jsonObject.Add("property", new JsonString(previousValue));
+
+            Assert.Equal(previousValue, ((JsonString)jsonObject["property"]).Value);
+
+            jsonObject.Add("property", new JsonString(newValue));
+
+            string expectedString = expected == ExpectedValue.Previous ? previousValue : newValue;
+            Assert.Equal(expectedString, ((JsonString) jsonObject["property"]).Value);
+
+            // with indexer, property should change no matter which duplicates handling option is chosen:
+            jsonObject["property"] = (JsonString)"indexer value";
+            Assert.Equal("indexer value", (JsonString)jsonObject["property"]);
+        }
+
+
+        [Theory]
+        [InlineData(DuplicatePropertyNameHandling.Replace, ExpectedValue.New)]
+        [InlineData(DuplicatePropertyNameHandling.Replace, ExpectedValue.New, false)]
+        [InlineData(DuplicatePropertyNameHandling.Ignore, ExpectedValue.Previous)]
+        public static void TestDuplicatesReplaceAndIgnore(DuplicatePropertyNameHandling duplicatePropertyNameHandling, ExpectedValue expected = default, bool valueConstructor = true)
```

Comment:
nit: The `ExpectedValue` enum seems unnecessary given the use and complicates the test logic a bit. Use a bool instead, or maybe even specify the expected string itself within the inline data.

Something like the following (and make the "default" ctor a separate test):
```C#
        [InlineData(DuplicatePropertyNameHandling.Replace, "value1", "value2", "value2")]
        [InlineData(DuplicatePropertyNameHandling.Ignore, "value1", "value2", "value1")]
```

Reply:
Also, why is the `expected` parameter optional?

If you plan on keeping the `valueConstructor` bool within the same test, rename to `useDefaultCtor`.

Reply:
I'll change those.